### PR TITLE
Move design to MVVM

### LINKED ISSUE #10: Add/Fix support for playing audio in the pregap
Currently, playing audio in the pregap doesn't seem to work, I'm not able to rewind to where it would be. May possibly be related to needing to fix reading negative sectors.

### LINKED ISSUE #22: [Refactor] Move UI to MVVM

### PR DESCRIPTION
**Pull request by mnadareski**
This PR has the overall goal of moving to an MVVM pattern (fixes #22).

As a part of this, the following changes are implemented:
- Removed all references to Settings object in anything but UI code
- Removed all knowledge of the UI from the view-model and model code
- Added subscriber events for model and view-model code
- Removed UI timer for forcing updates
- Provided more methods to safely set values going down the chain
- Separated out the GUI and common code into separate libraries for future use

Notably, on my development machine, this decreased CPU usage by a fair bit, especially at idle.

As a bonus, since it came up during the refactoring and testing, hidden tracks can now play properly and are controlled by an internal flag set by the settings with the current UI (fixes #10). This ended up being a pretty major thing to address as part of this PR, because a lot of the behavior that was causing hidden tracks to render incorrectly was tied to a poor implementation of track loading and switching.

### CODE REVIEW

**Comment by mnadareski** [comment]
Please note that I am opening this in a draft state because I am not 100% confident that I applied the MVVM pattern to the code correctly.

**Comment by mnadareski** [comment]
Marking this as ready to review finally, because I feel like I've isolated the pieces properly at this point. All of my local tests are working as expected, with no regressions that I have been able to spot.

**Comment by mnadareski** [comment]
All future work items have been added as appropriately-tagged issues. Nothing else will be added outside of requested changes.

File: RedBookPlayer.Common/Discs/CompactDisc.cs [hunk 1]
```diff
@@ -264,31 +418,63 @@ public override bool PreviousIndex(bool changeTrack)
...
-            CurrentTrackNumber = 0;
+            CurrentTrackNumber = 1;
             LoadTrack(CurrentTrackNumber);
         }
 
+        /// <summary>
+        /// Set the value for loading data tracks
+        /// </summary>
+        /// <param name="load">True to enable loading data tracks, false otherwise</param>
+        public void SetLoadDataTracks(bool load) => _loadDataTracks = load;
```
Comment (claunia) [inline RedBookPlayer.Common/Discs/CompactDisc.cs]: Use properties.

File: RedBookPlayer.Common/Discs/CompactDisc.cs [hunk 2]
```diff
@@ -264,31 +418,63 @@ public override bool PreviousIndex(bool changeTrack)
...
+
+        /// <summary>
+        /// Set the value for loading hidden tracks
+        /// </summary>
+        /// <param name="load">True to enable loading hidden tracks, false otherwise</param>
+        public void SetLoadHiddenTracks(bool load) => _loadHiddenTracks = load;
```
Comment (claunia) [inline RedBookPlayer.Common/Discs/CompactDisc.cs]: Use properties.

File: RedBookPlayer.Common/Discs/CompactDisc.cs [hunk 3]
```diff
@@ -6,11 +6,12 @@
 using Aaru.CommonTypes.Structs;
 using Aaru.Decoders.CD;
 using Aaru.Helpers;
+using ReactiveUI;
 using static Aaru.Decoders.CD.FullTOC;
 
-namespace RedBookPlayer.Discs
+namespace RedBookPlayer.Common.Discs
```
Comment (claunia) [inline RedBookPlayer.Common/Discs/CompactDisc.cs]: Is this a view, a model, or a viewmodel?

File: RedBookPlayer.Common/Discs/OpticalDisc.cs [hunk 4]
```diff
@@ -1,9 +1,10 @@
 using Aaru.CommonTypes.Enums;
 using Aaru.CommonTypes.Interfaces;
+using ReactiveUI;
 
-namespace RedBookPlayer.Discs
+namespace RedBookPlayer.Common.Discs
 {
-    public abstract class OpticalDisc
+    public abstract class OpticalDisc : ReactiveObject
```
Comment (claunia) [inline RedBookPlayer.Common/Discs/OpticalDisc.cs]: Shouldn't it be called OpticalDiscBase?

File: RedBookPlayer.Common/Discs/OpticalDisc.cs [hunk 5]
```diff
@@ -1,9 +1,10 @@
 using Aaru.CommonTypes.Enums;
 using Aaru.CommonTypes.Interfaces;
+using ReactiveUI;
 
-namespace RedBookPlayer.Discs
+namespace RedBookPlayer.Common.Discs
```
Comment (claunia) [inline RedBookPlayer.Common/Discs/OpticalDisc.cs]: View, Model or ViewModel?

File: RedBookPlayer.Common/Discs/OpticalDiscFactory.cs [hunk 6]
```diff
@@ -1,17 +1,57 @@
+using System.IO;
 using Aaru.CommonTypes.Interfaces;
 using Aaru.CommonTypes.Metadata;
+using Aaru.DiscImages;
+using Aaru.Filters;
 
-namespace RedBookPlayer.Discs
+namespace RedBookPlayer.Common.Discs
```
Comment (claunia) [inline RedBookPlayer.Common/Discs/OpticalDiscFactory.cs]: Factories should go in factories namespace and folder imo.

File: RedBookPlayer.Common/PlayerViewModel.cs [hunk 7]
```diff
@@ -0,0 +1,328 @@
+using System.ComponentModel;
+using ReactiveUI;
+using RedBookPlayer.Common.Hardware;
+
+namespace RedBookPlayer.Common
```
Comment (claunia) [inline RedBookPlayer.Common/PlayerViewModel.cs]: Namespace, namespace

File: RedBookPlayer.GUI/MainWindow.xaml.cs [hunk 8]
```diff
@@ -5,6 +5,7 @@
 using Avalonia.Controls;
 using Avalonia.Input;
 using Avalonia.Markup.Xaml;
+using RedBookPlayer.Common;
 
 namespace RedBookPlayer.GUI
```
Comment (claunia) [inline RedBookPlayer.GUI/MainWindow.xaml.cs]: namespace namespace. Views views.

File: RedBookPlayer.GUI/PlayerView.xaml.cs [hunk 9]
```diff
@@ -24,125 +25,162 @@ public class PlayerView : UserControl
...
+            InitializeDigits();
         }
 
+        #region Helpers
+
         /// <summary>
         /// Load an image from the path
         /// </summary>
         /// <param name="path">Path to the image to load</param>
         public async Task<bool> LoadImage(string path)
```
Comment (claunia) [inline RedBookPlayer.GUI/PlayerView.xaml.cs]: This should go in the viewmodel.

File: RedBookPlayer.GUI/PlayerView.xaml.cs [hunk 10]
```diff
@@ -24,125 +25,162 @@ public class PlayerView : UserControl
...
+            PlayerViewModel.SetLoadHiddenTracks(App.Settings.PlayHiddenTracks);
+        }
+
+        /// <summary>
+        /// Generate the digit string to be interpreted by the frontend
+        /// </summary>
+        /// <returns>String representing the digits for the frontend</returns>
+        private string GenerateDigitString()
+        {
+            // If the disc isn't initialized, return all '-' characters
```
Comment (claunia) [inline RedBookPlayer.GUI/PlayerView.xaml.cs]: This should go in the viewmodel.

File: RedBookPlayer.GUI/PlayerView.xaml.cs [hunk 11]
```diff
@@ -24,125 +25,162 @@ public class PlayerView : UserControl
...
-            _updateTimer = new Timer(1000 / 60);
+            return sectorTime;
+        }
 
-            _updateTimer.Elapsed += (sender, e) =>
+        /// <summary>
+        /// Generate a path selection dialog box
+        /// </summary>
+        /// <returns>User-selected path, if possible</returns>
+        private async Task<string> GetPath()
```
Comment (claunia) [inline RedBookPlayer.GUI/PlayerView.xaml.cs]: -> Viewmodel

File: RedBookPlayer.GUI/PlayerView.xaml.cs [hunk 12]
```diff
@@ -24,125 +25,162 @@ public class PlayerView : UserControl
...
         }
 
         /// <summary>
-        /// Initialize the UI based on the currently selected theme
+        /// Get current sector time, accounting for offsets
         /// </summary>
-        /// <param name="xaml">XAML data representing the theme, null for default</param>
-        private void InitializeComponent(string xaml)
+        /// <returns>ulong representing the current sector time</returns>
+        private ulong GetCurrentSectorTime()
```
Comment (claunia) [inline RedBookPlayer.GUI/PlayerView.xaml.cs]: -> Viewmodel

File: RedBookPlayer.GUI/PlayerView.xaml.cs [hunk 13]
```diff
@@ -181,20 +219,38 @@ private void InitializeDigits()
...
+            {
+                AvaloniaXamlLoader.Load(this);
+            }
+        }
+
+        /// <summary>
+        /// Update the UI from the view-model
         /// </summary>
-        private void UpdateView(object sender, ElapsedEventArgs e)
+        private void PlayerViewModelStateChanged(object sender, PropertyChangedEventArgs e)
```
Comment (claunia) [inline RedBookPlayer.GUI/PlayerView.xaml.cs]: -> Viewmodel

File: RedBookPlayer.GUI/PlayerView.xaml.cs [hunk 14]
```diff
@@ -211,13 +267,19 @@ public async void LoadButton_Click(object sender, RoutedEventArgs e)
...
+        public void PlayPauseButton_Click(object sender, RoutedEventArgs e)
+        {
+            if(PlayerViewModel.Playing == true)
+                PlayerViewModel.Pause();
+            else
+                PlayerViewModel.Play();
+        }
 
-        public void StopButton_Click(object sender, RoutedEventArgs e) => PlayerViewModel.Playing = null;
+        public void StopButton_Click(object sender, RoutedEventArgs e) => PlayerViewModel.Stop();
```
Comment (claunia) [inline RedBookPlayer.GUI/PlayerView.xaml.cs]: Binding

File: RedBookPlayer.GUI/PlayerView.xaml.cs [hunk 15]
```diff
@@ -211,13 +267,19 @@ public async void LoadButton_Click(object sender, RoutedEventArgs e)
...
         }
 
-        public void PlayButton_Click(object sender, RoutedEventArgs e) => PlayerViewModel.Playing = true;
+        public void PlayButton_Click(object sender, RoutedEventArgs e) => PlayerViewModel.Play();
 
-        public void PauseButton_Click(object sender, RoutedEventArgs e) => PlayerViewModel.Playing = false;
+        public void PauseButton_Click(object sender, RoutedEventArgs e) => PlayerViewModel.Pause();
 
-        public void PlayPauseButton_Click(object sender, RoutedEventArgs e) => PlayerViewModel.Playing = !(PlayerViewModel.Playing ?? false);
+        public void PlayPauseButton_Click(object sender, RoutedEventArgs e)
```
Comment (claunia) [inline RedBookPlayer.GUI/PlayerView.xaml.cs]: Binding

File: RedBookPlayer.GUI/PlayerView.xaml.cs [hunk 16]
```diff
@@ -211,13 +267,19 @@ public async void LoadButton_Click(object sender, RoutedEventArgs e)
             await LoadImage(path);
         }
 
-        public void PlayButton_Click(object sender, RoutedEventArgs e) => PlayerViewModel.Playing = true;
+        public void PlayButton_Click(object sender, RoutedEventArgs e) => PlayerViewModel.Play();
 
-        public void PauseButton_Click(object sender, RoutedEventArgs e) => PlayerViewModel.Playing = false;
+        public void PauseButton_Click(object sender, RoutedEventArgs e) => PlayerViewModel.Pause();
```
Comment (claunia) [inline RedBookPlayer.GUI/PlayerView.xaml.cs]: Binding

File: RedBookPlayer.GUI/PlayerView.xaml.cs [hunk 17]
```diff
@@ -211,13 +267,19 @@ public async void LoadButton_Click(object sender, RoutedEventArgs e)
             await LoadImage(path);
         }
 
-        public void PlayButton_Click(object sender, RoutedEventArgs e) => PlayerViewModel.Playing = true;
+        public void PlayButton_Click(object sender, RoutedEventArgs e) => PlayerViewModel.Play();
```
Comment (claunia) [inline RedBookPlayer.GUI/PlayerView.xaml.cs]: Binding

File: RedBookPlayer.GUI/PlayerView.xaml.cs [hunk 18]
```diff
@@ -231,17 +293,17 @@ public async void LoadButton_Click(object sender, RoutedEventArgs e)
 
         public void RewindButton_Click(object sender, RoutedEventArgs e) => PlayerViewModel.Rewind();
 
-        public void VolumeUpButton_Click(object sender, RoutedEventArgs e) => PlayerViewModel.Volume++;
+        public void VolumeUpButton_Click(object sender, RoutedEventArgs e) => PlayerViewModel.SetVolume(PlayerViewModel.Volume + 1);
```
Comment (claunia) [inline RedBookPlayer.GUI/PlayerView.xaml.cs]: Binding

File: RedBookPlayer.GUI/PlayerView.xaml.cs [hunk 19]
```diff
@@ -231,17 +293,17 @@ public async void LoadButton_Click(object sender, RoutedEventArgs e)
 
         public void RewindButton_Click(object sender, RoutedEventArgs e) => PlayerViewModel.Rewind();
```
Comment (claunia) [inline RedBookPlayer.GUI/PlayerView.xaml.cs]: Binding

File: RedBookPlayer.GUI/PlayerView.xaml.cs [hunk 20]
```diff
@@ -231,17 +293,17 @@ public async void LoadButton_Click(object sender, RoutedEventArgs e)
...
 
-        public void VolumeUpButton_Click(object sender, RoutedEventArgs e) => PlayerViewModel.Volume++;
+        public void VolumeUpButton_Click(object sender, RoutedEventArgs e) => PlayerViewModel.SetVolume(PlayerViewModel.Volume + 1);
 
-        public void VolumeDownButton_Click(object sender, RoutedEventArgs e) => PlayerViewModel.Volume--;
+        public void VolumeDownButton_Click(object sender, RoutedEventArgs e) => PlayerViewModel.SetVolume(PlayerViewModel.Volume - 1);
```
Comment (claunia) [inline RedBookPlayer.GUI/PlayerView.xaml.cs]: Binding

File: RedBookPlayer.GUI/PlayerView.xaml.cs [hunk 21]
```diff
@@ -231,17 +293,17 @@ public async void LoadButton_Click(object sender, RoutedEventArgs e)
...
 
         public void MuteToggleButton_Click(object sender, RoutedEventArgs e) => PlayerViewModel.ToggleMute();
 
-        public void EnableDeEmphasisButton_Click(object sender, RoutedEventArgs e) => PlayerViewModel.ApplyDeEmphasis = true;
+        public void EnableDeEmphasisButton_Click(object sender, RoutedEventArgs e) => PlayerViewModel.SetDeEmphasis(true);
```
Comment (claunia) [inline RedBookPlayer.GUI/PlayerView.xaml.cs]: Binding

File: RedBookPlayer.GUI/PlayerView.xaml.cs [hunk 22]
```diff
@@ -231,17 +293,17 @@ public async void LoadButton_Click(object sender, RoutedEventArgs e)
...
 
-        public void DisableDeEmphasisButton_Click(object sender, RoutedEventArgs e) => PlayerViewModel.ApplyDeEmphasis = false;
+        public void DisableDeEmphasisButton_Click(object sender, RoutedEventArgs e) => PlayerViewModel.SetDeEmphasis(false);
```
Comment (claunia) [inline RedBookPlayer.GUI/PlayerView.xaml.cs]: Binding

File: RedBookPlayer.GUI/PlayerView.xaml.cs [hunk 23]
```diff
@@ -231,17 +293,17 @@ public async void LoadButton_Click(object sender, RoutedEventArgs e)
...
 
-        public void EnableDisableDeEmphasisButton_Click(object sender, RoutedEventArgs e) => PlayerViewModel.ApplyDeEmphasis = !PlayerViewModel.ApplyDeEmphasis;
+        public void EnableDisableDeEmphasisButton_Click(object sender, RoutedEventArgs e) => PlayerViewModel.SetDeEmphasis(!PlayerViewModel.ApplyDeEmphasis);
```
Comment (claunia) [inline RedBookPlayer.GUI/PlayerView.xaml.cs]: Binding

**Review by claunia — CHANGES_REQUESTED** [review]

**Review by claunia — APPROVED** [review]

**Pull request merged**